src/main/kotlin/treeLib/bintrees/AVLTree.kt

```diff
@@ -0,0 +1,174 @@
+            if (node == null) {
+                return AVLNode(key, value)
+            }
+            if (key < node.key) {
+                node.left = addRec(node.left, key, value)
+            } else if (key > node.key) {
+                node.right = addRec(node.right, key, value)
+            }
+            return balanceNode(node)
+        }
+        if (root == null) {
+            root = AVLNode(key, value)
+            amountOfNodes += 1
+            return root
+        }
+        amountOfNodes += 1
+        return addRec(root, key, value)
+    }
+
+    private fun rotateLeft(nodeA: AVLNode<K, V>?): AVLNode<K, V>? {
+        if (nodeA != null) {
+            if (nodeA.right != null) {
+                val nodeB = nodeA.right
+                nodeA.right = nodeB?.left
+                nodeB?.left = nodeA
+                fixHeight(nodeA)
+                fixHeight(nodeB)
+                return nodeB
+            } else {
+                throw NullPointerException("Right child node cannot be null.")
+            }
+        } else {
+            throw NullPointerException("Node cannot be null.")
+        }
+    }
+
+    private fun rotateRight(nodeA: AVLNode<K, V>?): AVLNode<K, V>? {
+        if (nodeA != null) {
+            if (nodeA.left != null) {
+                val nodeB = nodeA.left
+                nodeA.left = nodeB?.right
+                nodeB?.right = nodeA
+                fixHeight(nodeA)
+                fixHeight(nodeB)
+                return nodeB
+            } else {
+                throw NullPointerException("Left child node cannot be null.")
+            }
+        } else {
+            throw NullPointerException("Node cannot be null.")
+        }
+    }
+
+    private fun height(node: AVLNode<K, V>?): Int {
+        if (node == null) {
+            return 0
+        }
+        return node.height
+    }
+
+    private fun balanceFactor(node: AVLNode<K, V>?): Int {
+        if (node == null) {
+            throw NullPointerException("Node cannot be null.")
+        }
+        return height(node.right) - height(node.left)
+    }
+
+    private fun fixHeight(node: AVLNode<K, V>?) {
+        if (node != null) {
+            val heightLeft = height(node.left)
+            val heightRight = height(node.right)
+            if (heightLeft > heightRight) {
+                node.height = heightLeft + 1
+            } else {
+                node.height = heightRight + 1
+            }
+        }
+    }
+
+    private fun balanceNode(node: AVLNode<K, V>?): AVLNode<K, V>? {
+        fixHeight(node)
+        if (balanceFactor(node) == 2) {
+            if (node != null) {
+                if (balanceFactor(node.right) < 0) {
+                    node.right = rotateRight(node.right)
+                }
+                if (node == root) {
+                    root = rotateLeft(node)
+                    return root
+                }
+                return rotateLeft(node)
+            }
+        }
```

Comment:
Вроде все классно, но есть один момент который было бы лучше поправить и это вложенность.Зачем писать два if если можно один! Примерно так:
```suggestion
        fixHeight(node)
        if (balanceFactor(node) == 2 && node != null) {
            if (balanceFactor(node.right) < 0) {
                node.right = rotateRight(node.right)
            }

            if (node == root) {
                root = rotateLeft(node)
                return root
            }

            return rotateLeft(node)
        }
```

Reply:
И правда. Красивее и компактнее.